obspy/io/sac/arrayio.py

```diff
@@ -0,0 +1,637 @@
+    #    list). That's a total of 632 bytes.
+    # --------------------------------------------------------------
+    hf = from_buffer(f.read(4 * 70), dtype=native_str(endian_str + 'f4'))
+    hi = from_buffer(f.read(4 * 40), dtype=native_str(endian_str + 'i4'))
+    hs = from_buffer(f.read(24 * 8), dtype=native_str('|S8'))
+
+    isVALID = is_valid_byteorder(hi)
+    if not isVALID:
+        if isSPECIFIED:
+            # specified but not valid. you dun messed up.
+            raise SacIOError("Incorrect byteorder {}".format(byteorder))
+        else:
+            # not valid, but not specified.
+            # swap the dtype interpretation (dtype.byteorder), but keep the
+            # bytes, so the arrays in memory reflect the bytes on disk
+            hf = hf.newbyteorder('S')
+            hi = hi.newbyteorder('S')
+
+    # we now have correct headers, let's use their correct byte order.
+    endian_str = hi.dtype.byteorder
+
+    # check header lengths
+    if len(hf) != 70 or len(hi) != 40 or len(hs) != 24:
+        hf = hi = hs = None
+        f.close()
+        raise SacIOError("Cannot read all header values")
+
+    npts = hi[HD.INTHDRS.index('npts')]
+
+    # check file size
+    if checksize:
+        cur_pos = f.tell()
+        length = f.seek(0, 2)
+        f.seek(cur_pos, 0)
+        if length != (632 + 4 * int(npts)):
+            msg = "File-size and theoretical size are inconsistent.\n" \
+                  "Check that headers are consistent with time series."
+            raise SacIOError(msg)
+
+
+    # --------------------------------------------------------------
+    # READ DATA
+    # --------------------------------------------------------------
+    if headonly:
+        data = None
+    else:
+        data = from_buffer(f.read(npts * 4), dtype=native_str(endian_str + 'f4'))
+
+        if len(data) != npts:
+            f.close()
+            raise SacIOError("Cannot read all data points")
+
+    if is_file_name:
+        f.close()
+
+    return hf, hi, hs, data
+
+
+def read_sac_ascii(source, headonly=False):
+    """
+    Read a SAC ASCII file.
+
+    Parameters
+    ----------
+    source : str for File-like object
+        Full path or File-like object from a SAC ASCII file on disk.
+    headonly : bool
+        If headonly is True, only read the header arrays not the data array.
+
+    Returns
+    -------
+    hf, hi, hs : numpy.ndarray
+        The float, integer, and string header arrays.
+    data : numpy.ndarray or None
+        float32 data array. If headonly is True, data will be None.
+
+    """
+    # checks: ASCII-ness, header array length, npts matches data length
+    try:
+        fh = open(source, 'rb')
+        is_file_name = True
+    except IOError:
+        raise SacIOError("No such file: " + source)
+    except TypeError:
+        fh = source
+        is_file_name = False
+
+    contents = fh.read()
+
+    contents = [_i.rstrip(b"\n\r") for _i in contents.splitlines(True)]
```

Comment:
This is an inherited snippet, and I don't know which edge case it was written for, it any.  Maybe I'll remove it in the current obspy master and see if any sacio/sac.core tests fail.